### PR TITLE
Enumerable#zip, Enumerable#each_with_index, Array#shift

### PR DESCRIPTION
**Pull request by alimpfard**
Progress towards #19.
Full support is blocked by #21, but I didn't feel like reading up on how exactly `Enumerator` is supposed to behave.

I'm not sure if the `Vector` changes go against the overall style of the project, please tell me if they do.

### CODE REVIEW

**Comment by alimpfard** [comment]
Some day I'll get used to the `#name` way of referring to messages in Ruby.
some day.

**Comment by seven1m** [comment]
This looks great! :heart_eyes: 

I did notice some failing specs when I ran the suite locally:

```
Enumerable#each_with_index
  provides each element to the block
    #<EnumerableSpecs::EachDefiner:0x7faef12c6680> should be == to nil (line 22)
Enumerable#each_with_index
  provides each element to the block and its index
    [2, 5, 3, 6, 1, 4] should be eql (same type) to #<EnumerableSpecs::Numerous:0x7faef12d6a90> (line 29)
Enumerable#each_with_index
  binds splat arguments properly
    [2, 5, 3, 6, 1, 4] should be eql (same type) to #<EnumerableSpecs::Numerous:0x7faef122f9c0> (line 36)
```

I don't know why GitHub isn't running the test suite on this PR, so that's weird... I'll look into that.

Also, your style is a bit different than mine. Next time I save these files clang-format is going to do its thing and change some style. Might be good if you could run clang-format so you get credit in the git history for all this amazing work. :wink:

**Comment by alimpfard** [comment]
> I did notice some failing specs when I ran the suite locally

Curious...they were passing yesterday...

> Also, your style is a bit different than mine

I tried staying close to what I saw, but as always managed to completely ignore the `.clang-format` file in the repo :P

**Comment by alimpfard** [comment]
That _should_ make it run on PRs as well, I think.

**Comment by alimpfard** [comment]
Ah, you've already committed the workflows thing, I'll drop that

**Comment by seven1m** [comment]
Tests are passing locally now! Not sure what changed, but :ok_hand:

**Comment by seven1m** [comment]
I ran `clang-format` locally and nothing changed. Merging!

File: src/enumerable.rb [hunk 1]
```diff
@@ -39,6 +39,21 @@ def any?(pattern = nil)
...
+  def each_with_index(*args)
+    index = 0
+    if block_given?
+      # FIXME: Expose ArgumentError.
+      raise "Wrong number of arguments, given #{args.size} expected 0" unless args.empty?
+      each do |item|
+        yield item, index
+        index += 1
+      end
+      self
```
Comment (alimpfard) [inline src/enumerable.rb:51]: This changed :)

**Pull request merged**